flocker/control/_persistence.py

```diff
@@ -21,16 +21,94 @@
 from twisted.internet.defer import succeed
 from twisted.internet.task import LoopingCall
 
-from ._model import SERIALIZABLE_CLASSES, Deployment
+from ._model import SERIALIZABLE_CLASSES, Deployment, Configuration
 
 
 # Serialization marker storing the class name:
 _CLASS_MARKER = u"$__class__$"
 
+# The latest configuration version. Configuration versions are
+# always integers.
+_CONFIG_VERSION = 2
 
-class _ConfigurationEncoder(JSONEncoder):
+# Map of serializable class names to classes
+_CONFIG_CLASS_MAP = {cls.__name__: cls for cls in SERIALIZABLE_CLASSES}
+
+
+class ConfigurationMigrationError(Exception):
+    """
+    Error raised when a configuration migration is unable to take place
+    or complete successfully.
+    """
+
+
+def migrate_configuration(source_version, target_version,
+                          config, migration_class):
+    """
+    Migrate a persisted configuration from one version to another
+    in sequential upgrades, e.g. a source version of 1 and target
+    version of 3 will perform two upgrades, from version 1 to 2,
+    followed by 2 to 3.
+
+    Calls the correct ``ConfigurationMigration`` class methods for
+    the suppled source and target versions.
+
+    :param int source_version: The version to migrate from.
+    :param int target_version: The version to migrate to.
+    :param bytes config: The JSON-encoded source configuration.
+    :param class migration_class: The class containing the methods
+        that will be used for migration.
+
+    :return bytes: The updated JSON configuration after migration.
+    """
+    upgraded_config = config
+    current_version = source_version
+    for upgrade_version in range(source_version + 1, target_version + 1):
+        with _LOG_UPGRADE(configuration=upgraded_config,
+                          source_version=current_version,
+                          target_version=upgrade_version):
+            migration_method = (
+                u"configuration_v%d_v%d"
```

Comment:
Since this driver only ever upgrades from N to N+1 the `v%d_v%d` mechanism - ie, upgrades between two arbitrary versions - seems like more than we need and is somewhat misleading (since `v1_v5` won't actually be used if it is defined).  Supporting that mechanism also introduces more complexity into the system (primarily once you have both `v1_v3` and `v1_v2` + `v2_v3` - even if you don't ever have them in the code base at the same time - since you can't be as sure how some state was upgraded anymore).  I'd suggest something like `upgrade_from_N` which is always used to convert state from `N` to `N + 1`.